### PR TITLE
CDAP-18332: Introduce retries to all non-mutation RemoteDatasetFramework methods.

### PR DESCRIPTION
**Pull request by tivv**

### CODE REVIEW

File: cdap-common/src/main/resources/cdap-default.xml [hunk 1]
```diff
@@ -1349,6 +1349,46 @@
     </description>
   </property>
 
+  <property>
+    <name>dataset.remote.retry.policy.base.delay.ms</name>
```
Comment (chtyim) [inline cdap-common/src/main/resources/cdap-default.xml]: Use `system.` prefix as a naming convention for retry policy.

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java [hunk 2]
```diff
@@ -158,37 +163,39 @@ public void updateInstance(DatasetId datasetInstanceId, DatasetProperties props)
...
     throws DatasetManagementException {
-    return clientCache.getUnchecked(namespaceId).getAllInstances();
+    return Retries.callWithRetries(() -> clientCache.getUnchecked(namespaceId).getAllInstances(), retryStrategy);
   }
 
   @Override
   public Collection<DatasetSpecificationSummary> getInstances(NamespaceId namespaceId, Map<String, String> properties)
     throws DatasetManagementException {
-    return clientCache.getUnchecked(namespaceId).getInstances(properties);
+    return Retries.callWithRetries(() -> clientCache.getUnchecked(namespaceId).getInstances(properties), retryStrategy);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: Does the `clientCache.getUnchecked` populate `RetryableException` correctly? If the getUnchecked wraps exception, then the `callWithRetries` may not work as intended.

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java [hunk 3]
```diff
@@ -158,37 +163,39 @@ public void updateInstance(DatasetId datasetInstanceId, DatasetProperties props)
...
   }
 
   @Nullable
   @Override
   public DatasetSpecification getDatasetSpec(DatasetId datasetInstanceId) throws DatasetManagementException {
```
Comment (janakigg) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java:183]: Are you planning to add unit tests?

Comment (tivv) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java:183]: Definitely not unit as mocking out all retry mechanism would provide little value, but having a small integration test for exception propagation would be good

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java (hunk 2 above)
Comment (tivv) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: done

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java (hunk 3 above)
Comment (tivv) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java:183]: done

**Comment by tivv** [comment]
@chtyim @janakigg please review again

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java [hunk 4]
```diff
@@ -79,10 +85,14 @@
 @SuppressWarnings("unchecked")
 public class RemoteDatasetFramework implements DatasetFramework {
   private static final Logger LOG = LoggerFactory.getLogger(RemoteDatasetFramework.class);
+  private static final Predicate<Throwable> RETRYABLE_PREDICATE = t ->
+    t instanceof RetryableException ||
+    (t  instanceof UncheckedExecutionException && t.getCause() instanceof RetryableException);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: Remove extra space after `t`

**Review by chtyim — APPROVED** [review]

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java (hunk 4 above)
Comment (tivv) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: Done

**Pull request merged**